flink-ml-lib/src/main/java/org/apache/flink/ml/classification/linear/LogisticRegression.java

```diff
@@ -0,0 +1,653 @@
+                                        double[] value = (double[]) x;
+                                        return value[value.length - 1] / value[value.length - 2];
+                                    })
+                            .flatMap(new TerminateOnMaxIterOrTol(maxIter, tol));
+
+            return new IterationBodyResult(
+                    DataStreamList.of(reducedGradientAndLoss.get(0)),
+                    DataStreamList.of(gradientAndLoss.getSideOutput(MODEL_OUTPUT)),
+                    terminationCriteria);
+        }
+    }
+
+    /**
+     * A stream operator that caches the training data in the first iteration and updates model
+     * using gradients iteratively. The first input is the training data, and the second input is
+     * the initialized model or feedback of gradient and loss.
+     */
+    private static class CacheDataAndUpdateModelAndComputeGradient
+            extends AbstractStreamOperator<double[]>
+            implements TwoInputStreamOperator<Tuple3<Double, Double, double[]>, double[], double[]>,
+                    IterationListener<double[]> {
+
+        double[] coefficient;
+
+        double[] gradient;
+
+        int dim;
+
+        int globalBatchSize;
+
+        int localBatchSize;
+
+        double learningRate;
+
+        int maxIteration;
+
+        double tol;
+
+        private final LogisticGradient logisticGradient;
+
+        /** TODO: add a more efficient sampling method. */
+        List<Tuple3<Double, Double, double[]>> cachedTrainData;
+
+        List<Tuple3<Double, Double, double[]>> batchData;
+
+        Random random = new Random(2021);
+
+        /** The buffer for feedback record: {coefficient, weightSum, loss}. */
+        double[] feedbackBuffer;
+
+        ListState<Double> lossState;
+
+        ListState<double[]> coefficientState;
+
+        ListState<Tuple3<Double, Double, double[]>> trainDataState;
+
+        ListState<double[]> feedbackBufferState;
+
+        public CacheDataAndUpdateModelAndComputeGradient(
+                LogisticGradient logisticGradient,
+                int globalBatchSize,
+                double learningRate,
+                int maxIteration,
+                double tol) {
+            this.logisticGradient = logisticGradient;
+            this.globalBatchSize = globalBatchSize;
+            this.batchData = new ArrayList<>(globalBatchSize);
+            this.learningRate = learningRate;
+            this.maxIteration = maxIteration;
+            this.tol = tol;
+        }
+
+        @Override
+        public void open() {
+            int numTasks = getRuntimeContext().getNumberOfParallelSubtasks();
+            int taskId = getRuntimeContext().getIndexOfThisSubtask();
+            localBatchSize = globalBatchSize / numTasks;
+            if (globalBatchSize % numTasks > taskId) {
+                localBatchSize++;
+            }
+        }
+
+        private List<Tuple3<Double, Double, double[]>> prepareBatchData(
```

Comment:
It does not seem common to use `prepare` in this case. Would it be more intuitive to name this e.g. `getRandomBatch()`?

Reply:
How about `getBatchData()`? The word `Batch` is usually enough in machine learning for this case.

Reply:
Sounds good.